### PR TITLE
[Trait] Metricstrait

### PR DESCRIPTION
**Pull request by ryanzhang-oss**
### Description of your changes
Remove the need to have a service by
- add a workloadRef to the metrics trait
- looking for a child service resource of the workload if there exists one
- create a service that selects the pods that belong to the workload (user needs to specify a label selector)

### How has this code been tested?
Write integration test using controller-runtime test framework (its delete function seems to be broken: delete logics that work with real K8s cluster stuck on the test framework. Have to find some workaround)

### Checklist

I have:
- [x] Title of the PR starts with OAM type (e.g. `[Workload]` or `[Trait]` or `[Scope]`).
- [ ] Updated/Added any relevant [documentation] and [examples].
- [x] Unit/E2E Tests added.

### CODE REVIEW

File: traits/metricstrait/config/samples/application/README.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
+# Deployment Workload
```
Comment (wonderflow) [inline traits/metricstrait/config/samples/application/README.md:1]: we should move this to  `traits/metricstrait/README.md`, or user can't view it in this place.

File: traits/metricstrait/hack/servcieMonitor.yaml [hunk 2]
```diff
@@ -0,0 +1,465 @@
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  annotations:
+    controller-gen.kubebuilder.io/version: v0.2.4
+  creationTimestamp: null
+  name: servicemonitors.monitoring.coreos.com
```
Comment (wonderflow) [inline traits/metricstrait/hack/servcieMonitor.yaml:7]: why put this CRD in hack? And I didn't see any where using it?

Comment (ryanzhang-oss) [reply traits/metricstrait/hack/servcieMonitor.yaml:7]: This is used in the controllers/suite_test.go, the test needs to be able to emit servcieMonitor using the mocking api-server

File: traits/metricstrait/config/samples/application/README.md (hunk 1 above)
Comment (ryanzhang-oss) [reply traits/metricstrait/config/samples/application/README.md:1]: it was designed for this example only and it is pretty much a "quick start". Maybe I can add a main  README that points to this? I can put more context in the main README.

Comment (wonderflow) [reply traits/metricstrait/config/samples/application/README.md:1]: Yes, add a  main  README that points to this is good!

File: traits/metricstrait/hack/servcieMonitor.yaml (hunk 2 above)
Comment (wonderflow) [reply traits/metricstrait/hack/servcieMonitor.yaml:7]: I see, maybe putting it to a folder under test is better?

**Pull request merged**